cli/src/main/java/keywhiz/cli/commands/DescribeAction.java

```diff
@@ -83,20 +84,49 @@ public DescribeAction(DescribeActionConfig describeActionConfig, KeywhizClient c
         break;
 
       case "secret":
-        SanitizedSecret sanitizedSecret;
-        try {
-          sanitizedSecret = keywhizClient.getSanitizedSecretByName(name);
-
-          printing.printSanitizedSecretWithDetails(sanitizedSecret);
-        } catch (NotFoundException e) {
-          throw new AssertionError("Secret not found.");
-        } catch (IOException e) {
-          throw Throwables.propagate(e);
-        }
+        describeSecret();
         break;
 
       default:
         throw new IllegalArgumentException("Invalid describe type specified: " + firstType);
     }
   }
+
+  private void describeSecret() {
+    SanitizedSecret sanitizedSecret = getSanitizedSecret();
+    List<SanitizedSecret> deletedSecrets = describeActionConfig.includeDeleted
+        ? getDeletedSecrets()
+        : null;
+
+    if (sanitizedSecret != null) {
+      printing.printSanitizedSecretWithDetails(sanitizedSecret);
+    }
+
+    if (deletedSecrets != null) {
+      printing.printDeletedSecretsWithDetails(deletedSecrets);
+    }
+  }
+
+  @Nullable
+  private SanitizedSecret getSanitizedSecret() {
+    try {
+      return keywhizClient.getSanitizedSecretByName(describeActionConfig.name);
+    } catch (NotFoundException e) {
+      if (!describeActionConfig.includeDeleted) {
+        throw new AssertionError("Secret not found.");
+      }
+      // If we're including deleted secrets, it's ok if no non-deleted secret was found
+      return null;
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
```

Comment:
Since the result is possibly null, we can better represent that by using an Optional<SanitizedSecret> and not have to worry about null checks. That also allows us to do something fancy in the calling method like

```
maybeSanitizedSecret.ifPresent(printing::printSanitizedSecretWithDetails);
```

The original code throws AssertionError, which is almost certainly wrong. I'd argue we just rethrow the NotFoundException as-is, but that might be annoying if it's checked (and also violates our current API contract).

Reply:
I agree throwing the NotFoundException as is makes sense. the worst case is that the CLI barfs when there's no secret found, which isn't that bad, right?